src/charm.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+#
+# Learn more at: https://juju.is/docs/sdk
+
+"""Charm the service."""
+
+import logging
+import socket
+from typing import Dict, List, Optional
+
+from charms.loki_k8s.v1.loki_push_api import _PebbleLogClient
+from cosl import JujuTopology
+from loki import (
+    CONTAINER_NAME,
+    LOKI_CERT_FILE,
+    Loki,
+)
+from loki_cluster import (
+    ConfigReceivedEvent,
+    LokiClusterRequirer,
+    LokiRole,
+)
+from ops.charm import CharmBase, CollectStatusEvent
+from ops.framework import BoundEvent, Object
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class LokiWorkerK8SOperatorCharm(CharmBase):
+    """A Juju Charmed Operator for Loki."""
+
+    _loki = None
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._container = self.unit.get_container(CONTAINER_NAME)
+        self._loki = Loki(self._container, self._loki_roles, self.model)
+        self._topology = JujuTopology.from_charm(self)
+        self._loki_cluster = LokiClusterRequirer(self)
+        self._log_forwarder = ManualLogForwarder(
+            charm=self,
+            loki_endpoints=self._loki_cluster.get_loki_endpoints(),
+            refresh_events=[
+                self.on["loki-cluster"].relation_joined,
+                self.on["loki-cluster"].relation_changed,
+            ],
+        )
+        self.unit.set_ports(self._loki.port)
+
+        # === EVENT HANDLER REGISTRATION === #
+        self.framework.observe(self.on.loki_pebble_ready, self._on_pebble_ready)  # pyright: ignore
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(
+            self._loki_cluster.on.config_received, self._on_loki_config_received
+        )
+        self.framework.observe(self._loki_cluster.on.created, self._on_loki_cluster_created)
```

Comment:
Do we need to directly observe here a relation managed by the the lib?
I imagine this is a tradeoff between this and emitting and observing custom events?

Reply:
Well, it is not a lib yet. We can refactor all this once we extract it to a lib.

We need to decide whether to place this `LokiClusterRequirer` and `LokiClusterProvider` classes. In the Coordinator? In the Worker? 🤔

Reply:
- Having it as a charm lib would we consistent with the approach in other charms. But perhaps you had a different plan for this one?
- The object already takes a charm instance: `self._loki_cluster = LokiClusterRequirer(self)`, so seems like the all `loki_cluster_*` relations can be observed there, and emit a custom event. To do that it doesn't have to be a charm lib.

Reply:
No, I hadn't.... my main doubt is where to place it.... Feels like the coordinator is the best place

Reply:
https://github.com/canonical/loki-worker-k8s-operator/issues/5